tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SeqOpsRule.scala

```diff
@@ -56,213 +58,222 @@ class SeqOpsRule(rewriter: SymbStateRewriter) extends RewritingRule {
-      val whenInRange2 =
-        tla.or(tla.not(inRange2) ? "b",
-            tla.eql(oracle.intCell.toNameEx ? "i", tla.plus(tla.int(i), offset2.toNameEx ? "i") ? "i") ? "b") ? "b"
-      val whenOutOfRange =
-        tla.or(tla.lt(tla.int(i), len1plus2.toNameEx ? "i") ? "b",
-            tla.eql(oracle.intCell.toNameEx ? "i", tla.int(ntotal)) ? "b") ? "b"
-
-      solverAssert(whenInRange1.typed(types, "b"))
-      solverAssert(whenInRange2.typed(types, "b"))
-      solverAssert(whenOutOfRange.typed(types, "b"))
+    // Create the proto sequence of the same capacity, which is the worst case scenario.
+    nextState = proto.make(nextState, capacity, copy)
+    val newProtoSeq = nextState.asCell
+
+    // newLen = IF adjustedEnd >= adjustedStart THEN 1 + adjustedEnd - adjustedStart ELSE 0
+    nextState = nextState.updateArena(_.appendCell(IntT()))
+    val newLen = nextState.arena.topCell
+    solverAssert(tla.eql(asInt(newLen),
+        tla.ite(tla.ge(asInt(adjustedEnd), asInt(adjustedStart)) as BoolT1(),
+            tla.plus(tla.int(1), tla.minus(asInt(adjustedEnd), asInt(adjustedStart)) as IntT1()) as IntT1(),
+            tla.int(0)) as IntT1()) as BoolT1())
+    proto.mkSeq(nextState, seqEx.typeTag.asTlaType1(), newProtoSeq, newLen)
+  }
 
-      pickedResult
+  /**
+   * Translate Append(s, e) which is canonically defined as:
+   *
+   * <pre> Append(s, e) == s \o << e >> </pre>
+   */
+  private def translateAppend(state: SymbState, seqEx: TlaEx, newElemEx: TlaEx): SymbState = {
+    // rewrite seqEx and newElemEx
+    var nextState = rewriter.rewriteUntilDone(state.setRex(seqEx))
+    val seqCell = nextState.asCell
+    val (protoSeq, len, capacity) = proto.unpackSeq(nextState.arena, seqCell)
+    nextState = rewriter.rewriteUntilDone(nextState.setRex(newElemEx))
+    val elemToAdd = nextState.asCell
+
+    def pick(state: SymbState, dstIndexBase1: Int): (SymbState, ArenaCell) = {
+      // Since we do not know the actual length of the sequence,
+      // we have to pick the value at every position, except the last one.
+      if (dstIndexBase1 > capacity) {
+        // the last element is for sure equal to newElem (if lenCell == capacity)
+        (state, elemToAdd)
+      } else {
+        // for all other elements, we pick either the element in the ith position, or the new element
+        val seqElem = proto.at(state.arena, protoSeq, dstIndexBase1)
+        val (oracleState, oracle) = picker.oracleFactory.newDefaultOracle(state, 2)
+        val newState = picker
+          .pickByOracle(oracleState, oracle, Seq(seqElem, elemToAdd), nextState.arena.cellTrue().toNameEx)
+        val pickedCell = newState.asCell
+        val cond = tla.le(tla.int(dstIndexBase1), len.toNameEx as IntT1()) as BoolT1()
+        val when0 = oracle.whenEqualTo(nextState, 0)
+        val ite = tla.ite(cond, when0, tla.not(when0) as BoolT1()) as BoolT1()
+        rewriter.solverContext.assertGroundExpr(ite)
```

Comment:
`IF p THEN q ELSE ~q` is just `p = q`, so 
```suggestion
        val eql = tla.eql(cond, when0 as BoolT1()) as BoolT1()
        rewriter.solverContext.assertGroundExpr(eql)
```